lib/internal/http2.js

```diff
@@ -48,7 +46,139 @@ Object.defineProperty(exports, 'constants', {
 // emit events as callbacks. To allow this to happen,
 // it has to inherit from EventEmitter.
 util.inherits(http2.Http2Session, EventEmitter);
+util.inherits(http2.Http2Stream, Duplex);
 
+function maybeDestroyStream(stream) {
+  if ((!stream.readable &&
+       !stream.writable &&
+       !stream._writableState.length)) {
+    stream.destroy();
+  }
+}
+
+http2.Http2Stream.prototype.onread = function onread(nread, buffer) {
+  const stream = this;
+
+  if (nread > 0) {
+    var ret = stream.push(buffer);
+    if (stream.reading && !ret) {
+      stream.reading = false;
+      var err = stream.readStop();
+      if (err) {
+        // TODO(jasnell): figure this out
+        maybeDestroyStream(stream);
+      }
+    }
+    return;
+  }
+
+  if (nread === 0) {
+    return;
+  }
+
+  if (nread !== uv.UV_EOF) {
+    // TODO(jasnell): figure out
+    return maybeDestroyStream(stream);
+  }
+
+  stream.push(null);
+
+  if (stream._readableState.length === 0) {
+    stream.readable = false;
+    // TODO(jasnell): Figure out
+    //maybeDestroy(self);
+  }
+};
+
+http2.Http2Stream.prototype.end = function(chunk, encoding, callback) {
+  Duplex.prototype.end.call(this, chunk, encoding, callback);
+  var req = new ShutdownWrap();
+  req.oncomplete = (status, handle, req) => {
+    process.nextTick(() => maybeDestroyStream(this));
+  };
```

Comment:
we can probably avoid this closure, by adding a reference to the http2stream to ShutdownWrap, and use a top level function.